### ISSUE TITLE
Tailscale client fails to connect when using mwan3 with static IP on primary WAN interface

### ISSUE THREAD

**Issue by qdrop17** [tailscale/tailscale#13357]
### What is the issue?

We are experiencing an issue with Tailscale connectivity when using OpenWRT as our gateway OS with mwan3 for multi-WAN setup. The problem occurs specifically when using a static IP on the primary WAN interface (wanA) while the secondary interface (wanB) uses a dynamic IP.

- Multi-WAN: Enabled using mwan3
- WAN Configuration: 
  - wanA: Static IP
  - wanB: Dynamic IP

When the above configuration is used, the Tailscale client fails to connect, resulting in the following error:

```
# Health check:
#     - not logged in, last login error=fetch control key: Get "https://controlplane.tailscale.com/key?v=85": dial tcp [2a05:d014:386:202:30f5:de62:8eed:5c47]:443: connect: network is unreachable

unexpected state: NoState
```

The issue only occurs with a static IP on wanA. Using dynamic IPs on both interfaces works correctly.
OpenWRT networking functions as expected:

```
root@gw-di-01:/# mwan3 status
Interface status:
 interface wan is offline 00h:00m:00s, uptime 00h:04m:23s and tracking is active
 interface wanb is online 00h:04m:03s, uptime 00h:04m:19s and tracking is active

root@gw-di-01:/# ping [google.ch](http://google.ch/)
PING [google.ch](http://google.ch/) (142.250.203.99): 56 data bytes
64 bytes from [142.250.203.99](http://142.250.203.99/): seq=0 ttl=111 time=3.034 ms
64 bytes from [142.250.203.99](http://142.250.203.99/): seq=1 ttl=111 time=2.983 ms
64 bytes from [142.250.203.99](http://142.250.203.99/): seq=2 ttl=111 time=3.019 ms
^C
--- [google.ch](http://google.ch/) ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 2.983/3.012/3.034 ms

root@gw-di-01:/# curl https://controlplane.tailscale.com/key?v=85
{
        "legacyPublicKey": "mkey:[redacted]",
        "publicKey": "mkey:[redacted]"
}
```

Removing the static IP route immediately resolves the Tailscale connectivity issue:
```
ip route del default via [redacted]
```

It appears that Tailscale is not respecting the host's networking configuration when a static IP is used on the primary WAN interface in conjunction with mwan3.

Related OpenWRT forum post: https://forum.openwrt.org/t/tailscale-and-mwan3-conflict-on-openwrt-connection-issues-and-routing-problems/205743 

Any help is greatly appreciated.

### Steps to reproduce

Using a OpenWRT gateway with 3 interfaces:

- WAN A (static IP)
- WAN B (DHCP)
- LAN

...and a default failover-configuration of mwan3 is enough to reproduce the issue.

It's also not relevant what Tailscale version we're using: The issue occurs up to 1.70.

### OS

Linux

### OS version

OpenWrt 23.05.2

### Tailscale version

1.58.2-1.70

**Comment by bradfitz** [comment]
Tailscale's apparently binding to the wrong device.

What does your route table look like normally, without Tailscale running, and without deleting the default route for Tailscale?

i.e. `ip r show table all`?

/cc @raggi @danderson

**Comment by qdrop17** [comment]
> Tailscale's apparently binding to the wrong device.
> 
> What does your route table look like normally, without Tailscale running, and without deleting the default route for Tailscale?
> 
> i.e. `ip r show table all`?
> 
> /cc @raggi @danderson

With Tailscale disabled:

```
root@gw-di-01:/# ip r show table all
default via 31.10.245.25 dev eth0 table 1 proto static metric 10 
31.10.245.24/30 dev eth0 table 1 proto static scope link metric 10 
172.22.64.0/22 dev eth2.42 table 1 proto kernel scope link src 172.22.64.1 
192.168.96.0/19 dev eth2.99 table 1 proto kernel scope link src 192.168.99.4 
192.168.224.0/19 dev eth2.43 table 1 proto kernel scope link src 192.168.224.1 
default via 172.21.32.1 dev eth1 table 2 proto static src 172.21.60.147 metric 20 
172.21.32.0/19 dev eth1 table 2 proto static scope link metric 20 
172.22.64.0/22 dev eth2.42 table 2 proto kernel scope link src 172.22.64.1 
192.168.96.0/19 dev eth2.99 table 2 proto kernel scope link src 192.168.99.4 
192.168.224.0/19 dev eth2.43 table 2 proto kernel scope link src 192.168.224.1 
default via 31.10.245.25 dev eth0 proto static metric 10 
default via 172.21.32.1 dev eth1 proto static src 172.21.60.147 metric 20 
31.10.245.24/30 dev eth0 proto static scope link metric 10 
172.21.32.0/19 dev eth1 proto static scope link metric 20 
172.22.64.0/22 dev eth2.42 proto kernel scope link src 172.22.64.1 
192.168.96.0/19 dev eth2.99 proto kernel scope link src 192.168.99.4 
192.168.224.0/19 dev eth2.43 proto kernel scope link src 192.168.224.1 
local 31.10.245.26 dev eth0 table local proto kernel scope host src 31.10.245.26 
broadcast 31.10.245.27 dev eth0 table local proto kernel scope link src 31.10.245.26 
local 127.0.0.0/8 dev lo table local proto kernel scope host src 127.0.0.1 
local 127.0.0.1 dev lo table local proto kernel scope host src 127.0.0.1 
broadcast 127.255.255.255 dev lo table local proto kernel scope link src 127.0.0.1 
local 172.21.60.147 dev eth1 table local proto kernel scope host src 172.21.60.147 
broadcast 172.21.63.255 dev eth1 table local proto kernel scope link src 172.21.60.147 
local 172.22.64.1 dev eth2.42 table local proto kernel scope host src 172.22.64.1 
broadcast 172.22.67.255 dev eth2.42 table local proto kernel scope link src 172.22.64.1 
local 192.168.99.4 dev eth2.99 table local proto kernel scope host src 192.168.99.4 
broadcast 192.168.127.255 dev eth2.99 table local proto kernel scope link src 192.168.99.4 
local 192.168.224.1 dev eth2.43 table local proto kernel scope host src 192.168.224.1 
broadcast 192.168.255.255 dev eth2.43 table local proto kernel scope link src 192.168.224.1 
fe80::/64 dev eth2 proto kernel metric 256 pref medium
fe80::/64 dev eth1 proto kernel metric 256 pref medium
local ::1 dev lo table local proto kernel metric 0 pref medium
anycast fe80:: dev eth2 table local proto kernel metric 0 pref medium
anycast fe80:: dev eth1 table local proto kernel metric 0 pref medium
local fe80::be24:11ff:fe47:5b8a dev eth1 table local proto kernel metric 0 pref medium
local fe80::be24:11ff:febc:726e dev eth2 table local proto kernel metric 0 pref medium
multicast ff00::/8 dev eth2 table local proto kernel metric 256 pref medium
multicast ff00::/8 dev eth1 table local proto kernel metric 256 pref medium
```

...with Tailscale enabled:

```
root@gw-di-01:/# ip r show table all
default via 31.10.245.25 dev eth0 table 1 proto static metric 10 
31.10.245.24/30 dev eth0 table 1 proto static scope link metric 10 
172.22.64.0/22 dev eth2.42 table 1 proto kernel scope link src 172.22.64.1 
192.168.96.0/19 dev eth2.99 table 1 proto kernel scope link src 192.168.99.4 
192.168.224.0/19 dev eth2.43 table 1 proto kernel scope link src 192.168.224.1 
default via 172.21.32.1 dev eth1 table 2 proto static src 172.21.60.147 metric 20 
172.21.32.0/19 dev eth1 table 2 proto static scope link metric 20 
172.22.64.0/22 dev eth2.42 table 2 proto kernel scope link src 172.22.64.1 
192.168.96.0/19 dev eth2.99 table 2 proto kernel scope link src 192.168.99.4 
192.168.224.0/19 dev eth2.43 table 2 proto kernel scope link src 192.168.224.1 
default via 31.10.245.25 dev eth0 proto static metric 10 
default via 172.21.32.1 dev eth1 proto static src 172.21.60.147 metric 20 
31.10.245.24/30 dev eth0 proto static scope link metric 10 
172.21.32.0/19 dev eth1 proto static scope link metric 20 
172.22.64.0/22 dev eth2.42 proto kernel scope link src 172.22.64.1 
192.168.96.0/19 dev eth2.99 proto kernel scope link src 192.168.99.4 
192.168.224.0/19 dev eth2.43 proto kernel scope link src 192.168.224.1 
local 31.10.245.26 dev eth0 table local proto kernel scope host src 31.10.245.26 
broadcast 31.10.245.27 dev eth0 table local proto kernel scope link src 31.10.245.26 
local 127.0.0.0/8 dev lo table local proto kernel scope host src 127.0.0.1 
local 127.0.0.1 dev lo table local proto kernel scope host src 127.0.0.1 
broadcast 127.255.255.255 dev lo table local proto kernel scope link src 127.0.0.1 
local 172.21.60.147 dev eth1 table local proto kernel scope host src 172.21.60.147 
broadcast 172.21.63.255 dev eth1 table local proto kernel scope link src 172.21.60.147 
local 172.22.64.1 dev eth2.42 table local proto kernel scope host src 172.22.64.1 
broadcast 172.22.67.255 dev eth2.42 table local proto kernel scope link src 172.22.64.1 
local 192.168.99.4 dev eth2.99 table local proto kernel scope host src 192.168.99.4 
broadcast 192.168.127.255 dev eth2.99 table local proto kernel scope link src 192.168.99.4 
local 192.168.224.1 dev eth2.43 table local proto kernel scope host src 192.168.224.1 
broadcast 192.168.255.255 dev eth2.43 table local proto kernel scope link src 192.168.224.1 
fe80::/64 dev eth2 proto kernel metric 256 pref medium
fe80::/64 dev eth1 proto kernel metric 256 pref medium
fe80::/64 dev tailscale0 proto kernel metric 256 pref medium
local ::1 dev lo table local proto kernel metric 0 pref medium
anycast fe80:: dev eth2 table local proto kernel metric 0 pref medium
anycast fe80:: dev eth1 table local proto kernel metric 0 pref medium
anycast fe80:: dev tailscale0 table local proto kernel metric 0 pref medium
local fe80::be24:11ff:fe47:5b8a dev eth1 table local proto kernel metric 0 pref medium
local fe80::be24:11ff:febc:726e dev eth2 table local proto kernel metric 0 pref medium
local fe80::d2da:b582:e67a:eaac dev tailscale0 table local proto kernel metric 0 pref medium
multicast ff00::/8 dev eth2 table local proto kernel metric 256 pref medium
multicast ff00::/8 dev eth1 table local proto kernel metric 256 pref medium
multicast ff00::/8 dev tailscale0 table local proto kernel metric 256 pref medium
```

I hope that helps. Please don't hesitate to reach out if you need further assistance in pinpointing the issue.

**Comment by raggi** [comment]
the udp bind is separate from the control plane connection, which is what was failing in the health check.

```
default via 31.10.245.25 dev eth0 proto static metric 10 
```

> In the logs we even see how Tailscale specifically selects the wrong interface as defaultRoute:
> Fri Sep  6 14:46:25 2024 daemon.err tailscaled[4363]: 2024/09/06 12:46:25 LinkChange: major, rebinding. New state: interfaces.State{defaultRoute=eth0 ifs={eth0:[31.10.245.26/30] eth1:[172.21.44.66/19 llu6] eth2.42:[172.22.64.1/22] eth2.43:[192.168.224.1/19] eth2.99:[192.168.99.4/19]} v4=true v6=false}

You say `eth0` is the wrong interface here, why? It's the lowest metric interface in the first routing table. The interface appears to still be up and configured.

How is your failover configured?

**Comment by qdrop17** [comment]
After further investigation, we believe we've identified the root cause of the issue. Tailscale relies on the "main" routing table, which becomes dysfunctional during a failover event with a static IP. Unlike DHCP, where the route is discarded along with the invalid lease, the route persists with a static IP.

We've found that manually deleting the faulty default route in the "main" routing table (via /etc/mwan3.user) allows Tailscale to immediately reconnect using wanB. However, this approach breaks mwan3 functionality, as mwan3track stops working correctly.

While we can restore wanA by manually restarting mwan3 (/etc/init.d/mwan restart), we haven't yet found a satisfactory automatic workaround that doesn't involve completely managing the routing table through custom shell scripts.

We're seeking advice on how to achieve compatibility between Tailscale and mwan3. We suspect this might be related to the fwmarks used by mwan3. Any insights on how to resolve this conflict or suggestions for a more pragmatic solution that doesn't deviate significantly from the default configuration would be greatly appreciated.

**Comment by junx964** [comment]
What is the current situation? Is there a solution?

**Comment by junx964** [comment]
I saw the following website description:
https://tailscale.com/kb/1023/troubleshooting#lan-traffic-prioritization-with-overlapping-subnet-routes
```json
On OpenWRT systems detected as running mwan3, Tailscale rules are installed at a lower priority for compatibility reasons. On such systems, ip rules are installed with priorities ranging 1300—1400 instead of 5200—5300.
```

My current approach is to add the following to the startup section of /etc/init.d/tailscale:
```json
while true; do
    if /sbin/ip rule | grep '0x300/0x3f00' -q; then
        break
    fi
    sleep 1
done
```
Very low, but it works. This method only applies to devices with mwan3.

**Comment by qdrop17** [comment]
We are currently using a workaround through the mwan3.user file to manage this issue. This solution involves dynamically modifying the main routing table and adding a new route with a lower metric to force traffic through the functioning interface:

```
#!/bin/sh
if [ "${ACTION}" = "disconnected" ] && [ "${INTERFACE}" = "wan" ] ; then
    # When WAN down - set WANb as primary
    ip route del default metric 5 2>/dev/null
    ip route add default via $(ip route show dev eth1 | awk '/default/ {print $3}') dev eth1 metric 5
fi

if [ "${ACTION}" = "connected" ] && [ "${INTERFACE}" = "wan" ] ; then
    # When WAN up - remove WANb priority route
    ip route del default metric 5 2>/dev/null
fi

if [ "${ACTION}" = "disconnected" ] && [ "${INTERFACE}" = "wanb" ] ; then
    # When WANb down - set WAN as primary
    ip route del default metric 5 2>/dev/null
    ip route add default via $(ip route show dev eth0 | awk '/default/ {print $3}') dev eth0 metric 5
fi

if [ "${ACTION}" = "connected" ] && [ "${INTERFACE}" = "wanb" ] ; then
    # When WANb up - remove WAN priority route
    ip route del default metric 5 2>/dev/null
fi
```

However, this approach compromises several core mwan3 features, particularly the ability to load-balance traffic between multiple WAN connections.

We'll test @junx964 suggested workaround as an alternative solution. Thank you for the recommendation!

**Comment by junx964** [comment]
> We are currently using a workaround through the mwan3.user file to manage this issue. This solution involves dynamically modifying the main routing table and adding a new route with a lower metric to force traffic through the functioning interface:
> 
> ```
> #!/bin/sh
> if [ "${ACTION}" = "disconnected" ] && [ "${INTERFACE}" = "wan" ] ; then
>     # When WAN down - set WANb as primary
>     ip route del default metric 5 2>/dev/null
>     ip route add default via $(ip route show dev eth1 | awk '/default/ {print $3}') dev eth1 metric 5
> fi
> 
> if [ "${ACTION}" = "connected" ] && [ "${INTERFACE}" = "wan" ] ; then
>     # When WAN up - remove WANb priority route
>     ip route del default metric 5 2>/dev/null
> fi
> 
> if [ "${ACTION}" = "disconnected" ] && [ "${INTERFACE}" = "wanb" ] ; then
>     # When WANb down - set WAN as primary
>     ip route del default metric 5 2>/dev/null
>     ip route add default via $(ip route show dev eth0 | awk '/default/ {print $3}') dev eth0 metric 5
> fi
> 
> if [ "${ACTION}" = "connected" ] && [ "${INTERFACE}" = "wanb" ] ; then
>     # When WANb up - remove WAN priority route
>     ip route del default metric 5 2>/dev/null
> fi
> ```
> 
> However, this approach compromises several core mwan3 features, particularly the ability to load-balance traffic between multiple WAN connections.
> 
> We'll test @junx964's workaround, though we doubt it will solve our issue. While adding a longer sleep during Tailscale initialization helped with the first start, connections still drop during failover on static IP setups.


As long as tailscale is started after mwan3 configures the routing rule, it will be fine.

When the tailscale daemon is started, mwan3 is running and the routing rule is configured. At this time, the priority of tailscale's routing rule will be 1300-1400

If tailscale is not restarted, the tailscale routing rule will not change. No matter how mwan3 is restarted, it will not affect the tailscale routing rule.

**Comment by junx964** [comment]
I don't know if there is any software similar to mwan3. If there is, every time the user restarts, he needs to log in to the routing system to restart tailscale or check whether software similar to mwan3 is started and configure the routing rule before starting tailscale. I feel it is unreasonable.

Putting the priority of the routing rule to 1300-1400 by restarting tailscale is like a patch repair solution. It does not fundamentally solve the problem. I hope tailscale can optimize this problem.

Not only the routing system, if the client has multiple WANs and similar software is installed, this problem will also exist.
@raggi

**Comment by qdrop17** [comment]
> > We are currently using a workaround through the mwan3.user file to manage this issue. This solution involves dynamically modifying the main routing table and adding a new route with a lower metric to force traffic through the functioning interface:
> > ```
> > #!/bin/sh
> > if [ "${ACTION}" = "disconnected" ] && [ "${INTERFACE}" = "wan" ] ; then
> >     # When WAN down - set WANb as primary
> >     ip route del default metric 5 2>/dev/null
> >     ip route add default via $(ip route show dev eth1 | awk '/default/ {print $3}') dev eth1 metric 5
> > fi
> > 
> > if [ "${ACTION}" = "connected" ] && [ "${INTERFACE}" = "wan" ] ; then
> >     # When WAN up - remove WANb priority route
> >     ip route del default metric 5 2>/dev/null
> > fi
> > 
> > if [ "${ACTION}" = "disconnected" ] && [ "${INTERFACE}" = "wanb" ] ; then
> >     # When WANb down - set WAN as primary
> >     ip route del default metric 5 2>/dev/null
> >     ip route add default via $(ip route show dev eth0 | awk '/default/ {print $3}') dev eth0 metric 5
> > fi
> > 
> > if [ "${ACTION}" = "connected" ] && [ "${INTERFACE}" = "wanb" ] ; then
> >     # When WANb up - remove WAN priority route
> >     ip route del default metric 5 2>/dev/null
> > fi
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > However, this approach compromises several core mwan3 features, particularly the ability to load-balance traffic between multiple WAN connections.
> > We'll test @junx964's workaround, though we doubt it will solve our issue. While adding a longer sleep during Tailscale initialization helped with the first start, connections still drop during failover on static IP setups.
> 
> As long as tailscale is started after mwan3 configures the routing rule, it will be fine.
> 
> When the tailscale daemon is started, mwan3 is running and the routing rule is configured. At this time, the priority of tailscale's routing rule will be 1300-1400
> 
> If tailscale is not restarted, the tailscale routing rule will not change. No matter how mwan3 is restarted, it will not affect the tailscale routing rule.

We've thoroughly tested this issue and can confirm the following:

1. The problem specifically occurs when using static IPs on WAN interfaces.
2. Various timing adjustments (delayed starts/restarts of Tailscale) did not resolve the issue.
3. During failover events, Tailscale completely loses internet connectivity, including:
   - Unable to reach control plane servers
   - Cannot establish VPN tunnels
   - Matches the behavior described in the original issue

Notably, while Tailscale connectivity fails, both the OpenWRT router and the internal subnet continue to function normally. This suggests the issue is specific to Tailscale's handling of the failover event together with the mwan3 architecture.

**Comment by qdrop17** [comment]
> @qdrop17 Why do I feel like your issue isn't really related to mwan3?
> 
> Could it be that your static WAN interface is unable to connect to the internet?
> 
> You can specify the interface to test when pinging:
> 
> ```shell
> ping [google.ch](http://google.ch/) -I <interface name>
> ```

The issue is indeed related to mwan3, though not necessarily caused by it. The problem likely stems from the interaction between mwan3 and Tailscale. As previously noted, mwan3 functions correctly with a generic configuration, maintaining connectivity for connected subnets regardless of the active interface. The problem specifically affects Tailscale and only occurs when using static IPs on WAN interfaces.

We have been successful in reproducing it.

**Comment by bassopt** [comment]
Hello.
Was this ever solved?
I'm currently using mwan3 LTE modem and FTTH as failover, both with dynamic IP and I have the same problem.
The way i could mend it somehow was to add a sleep 20 at init.d tailscale but it's a bit annoying since it doesn't surivice openwrt upgrades.

**Comment by qdrop17** [comment]
We've migrated to a custom-built failover script. It significantly reduces complexity and ensures Tailscale works flawlessly. While our script handles complete traffic failover, it doesn't support load balancing—though we don't require that functionality.

**Comment by e-minguez** [comment]
> We've migrated to a custom-built failover script. It significantly reduces complexity and ensures Tailscale works flawlessly. While our script handles complete traffic failover, it doesn't support load balancing—though we don't require that functionality.

Would you mind sharing it? 😁

**Comment by ccadruvi** [comment]
@e-minguez I'm working in the same team as @qdrop17 , you can find the script here:
https://github.com/DigitecGalaxus/ansible-openwrt-framework/blob/master/roles/openwrt/wan_failover/files/wan-failover.sh

**Comment by e-minguez** [comment]
> [@e-minguez](https://github.com/e-minguez) I'm working in the same team as [@qdrop17](https://github.com/qdrop17) , you can find the script here: https://github.com/DigitecGalaxus/mesh-network/blob/master/wan_failover/files/wan-failover.sh

Thanks. I've finally managed to have some availability and replaced mwan3 with your script there.

Just in case, as you don't have issues enabled there I've seen `eth1` is hardcoded https://github.com/DigitecGalaxus/mesh-network/blob/master/wan_failover/files/wan-failover.sh#L133 in case you want to fix it.

**Comment by andriej** [comment]
And the script is gone..

**Comment by ccadruvi** [comment]
> And the script is gone..

We repurposed the repository and changed it's name. It's now here: https://github.com/DigitecGalaxus/ansible-openwrt-framework/blob/master/roles/openwrt/wan_failover/files/wan-failover.sh

**Comment by dhanuhammer** [comment]
I had a similar problem. 

I solved it with just changing metrics.
I use 3 uplinks from ISP A (assume ppoe1, ppoe2, static1), 1 uplink from ISP B (static2), so if all links from isp A go down, only then ISP B is used. The problem I had is, since there is a static IP config from ISP A, even if the internet is switched to ISP B, tailscale does not switch over. 

After a lot of searching and torturing Perplexity, I did one simple trick: I set the metric of ISP A's static interface to be lower than ISP B. So previously the metric was 1,2,3 for ppoe1, ppoe2, static1 (from ISP A) and 4 for static2 from ISP B. 

Now I changed it to 1,2,4 for ppoe1, ppoe2, static1 (from ISP A), 3 for static2 (from ISP B). when ISP A is up, all the 3 uplinks work together irrespective of the metric so I get load balancing. if ISP A goes down for eg., due to a fiber cut, the whole traffic shifts to ISP B (which is a 5G WAN). 

Tailscale used to prefer the static interface from ISP A over ISP B because its metric was higher, but now that its metric is lower, tailscale prefers the ISP B over A, even if ISP A' s static Interface still exists in the routing table. Once ISP A is back, tailscale is normally routed thru ISP A.
